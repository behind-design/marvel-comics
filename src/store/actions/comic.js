export const FETCH_COMIC_REQUEST = '@@marvel-comics/FETCH_COMIC_REQUEST';
export const FETCH_COMIC_SUCCESS = '@@marvel-comics/FETCH_COMIC_SUCCESS';
export const FETCH_COMIC_FAILURE = '@@marvel-comics/FETCH_COMIC_FAILURE';

export const fetchComic = comicId => ({
  type: FETCH_COMIC_REQUEST,
  comicId
});

export const fetchComicSuccess = (comicId, response) => ({
  type: FETCH_COMIC_SUCCESS,
  response,
  comicId
});

export const fetchComicFailure = (comicId, message) => ({
  type: FETCH_COMIC_FAILURE,
  message,
  comicId,
});
