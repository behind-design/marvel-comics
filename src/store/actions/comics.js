export const FETCH_COMICS_REQUEST = '@@marvel-comics/FETCH_COMICS_REQUEST';
export const FETCH_COMICS_SUCCESS = '@@marvel-comics/FETCH_COMICS_SUCCESS';
export const FETCH_COMICS_FAILURE = '@@marvel-comics/FETCH_COMICS_FAILURE';

export const fetchComics = (offset = 0) => ({
  type: FETCH_COMICS_REQUEST,
  offset
});

export const fetchComicsSuccess = ({ response, offset, total }) => ({
  type: FETCH_COMICS_SUCCESS,
  response,
  offset,
  total
});

export const fetchComicsFailure = message => ({
  type: FETCH_COMICS_FAILURE,
  message,
});
