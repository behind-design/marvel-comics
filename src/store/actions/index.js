import {
  FETCH_COMICS_REQUEST,
  FETCH_COMICS_SUCCESS,
  FETCH_COMICS_FAILURE,
  fetchComics,
  fetchComicsSuccess,
  fetchComicsFailure
} from './comics';
import {
  FETCH_COMIC_REQUEST,
  FETCH_COMIC_SUCCESS,
  FETCH_COMIC_FAILURE,
  fetchComic,
  fetchComicSuccess,
  fetchComicFailure
} from './comic';


const LOCATION_CHANGE = "@@router/LOCATION_CHANGE";
const CANCEL_REQUESTS = "@@marvel-comics/CANCEL_REQUESTS";

const cancelRequests = () => ({
  type: CANCEL_REQUESTS,
});

export {
  // Action types
  FETCH_COMICS_REQUEST,
  FETCH_COMICS_SUCCESS,
  FETCH_COMICS_FAILURE,

  FETCH_COMIC_REQUEST,
  FETCH_COMIC_SUCCESS,
  FETCH_COMIC_FAILURE,

  CANCEL_REQUESTS,
  LOCATION_CHANGE,

  // Action creators
  cancelRequests,

  fetchComics,
  fetchComicsSuccess,
  fetchComicsFailure,

  fetchComic,
  fetchComicSuccess,
  fetchComicFailure,
}
