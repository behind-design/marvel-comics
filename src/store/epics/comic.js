import { Observable } from 'rxjs/Rx';
import { normalize } from 'normalizr';

import cancelActions from './utils/cancelActions';
import ajaxUntilTotal from './utils/ajaxUntilTotal';
import * as schema from '../schema';
import {
  LOCATION_CHANGE,
  CANCEL_REQUESTS,
  FETCH_COMIC_REQUEST,
  fetchComicSuccess,
  fetchComicFailure
} from '../actions';


const requests = [
  [':id', {}],
  [':id/creators', { orderBy: 'lastName' }],
  [':id/characters', { orderBy: 'name' }],
];

const comicError = (comicId) => (error) =>
  Observable.of(fetchComicFailure(comicId, error.message || 'Error Fetching Comic Detail'));

const comicSuccess = (comicId) => ([comicResponse, creatorsResponse, charactersResponse]) => {
  const { entities: { comics }} = normalize(comicResponse[0], schema.comic);
  return fetchComicSuccess(comicId, {
    [comicId]: {
      ...comics[comicId],
      creators: [...creatorsResponse],
      characters: [...charactersResponse],
    }
  });
}

/**
 * This epic will fetch "concurrently" the following items:
 *  - The comic detail
 *  - The comic creators
 *  - The comic characters
 *
 *  Because of some limitations of the api when calling the
 *  "creators" and "characters" endpoints, we can encounter that we don't get back
 *  all the "results" available in the api. Because of this, we perform as
 *  many ajax request as we need until we have all the necessary data to merge
 *  back into the "original comic" and then fire the "FETCH_COMIC_SUCCESS" action.
 */
export const fetchComicEpic = (action$, store, { ajax }) =>
  action$
    .ofType(FETCH_COMIC_REQUEST)
    .switchMap(({ comicId }) =>
      Observable.forkJoin(requests.map(ajaxUntilTotal(comicId, ajax)))
        .takeUntil(cancelActions(action$, CANCEL_REQUESTS, LOCATION_CHANGE))
        .map(comicSuccess(comicId))
        .catch(comicError(comicId))
    );
