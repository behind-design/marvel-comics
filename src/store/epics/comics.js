import { Observable } from 'rxjs/Rx';
import { normalize } from 'normalizr';

import buildUrl from './utils/buildUrl';
import cancelActions from './utils/cancelActions';
import * as schema from '../schema';
import {
  CANCEL_REQUESTS,
  LOCATION_CHANGE,
  FETCH_COMICS_REQUEST,
  fetchComicsSuccess,
  fetchComicsFailure
} from '../actions';


const comicsRequest = (offset = 0) => ({
  method: 'GET',
  crossDomain: true,
  url: buildUrl('/comics', {
    offset,
    limit: 25,
    format: 'comic',
    noVariants: true,
    formatType: 'comic',
    orderBy: 'onsaleDate',
  })
});

const comicsError = (error) =>
  Observable.of(fetchComicsFailure(error.message || 'Error Fetching Comics List'));

const comicsSuccess = (action) => ({ results = [], total = 0 }) => {
  const offset = action.offset + results.length;
  return fetchComicsSuccess({
    response: normalize(results, schema.arrayOfComics),
    offset,
    total,
  });
};

export const fetchComicsEpic = (action$, store, { ajax }) =>
    action$
      .ofType(FETCH_COMICS_REQUEST)
      .switchMap(action =>
        ajax(comicsRequest(action.offset))
          .pluck('response').pluck('data')
          .takeUntil(cancelActions(action$, LOCATION_CHANGE, CANCEL_REQUESTS))
          .map(comicsSuccess(action))
          .catch(comicsError)
      );
