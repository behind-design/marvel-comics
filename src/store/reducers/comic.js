import { combineReducers } from "redux";

import createErrorReducer from "./createErrorReducer";
import createFetchingReducer from "./createFetchingReducer";
import {
  LOCATION_CHANGE,
  FETCH_COMIC_REQUEST,
  FETCH_COMIC_SUCCESS,
  FETCH_COMIC_FAILURE
} from "../actions/index";


export const detail = (state = null, action) => {
  switch (action.type) {
    case FETCH_COMIC_SUCCESS:
      return action.response;
    case LOCATION_CHANGE:
    case FETCH_COMIC_REQUEST:
    case FETCH_COMIC_FAILURE:
      return null;
    default:
      return state;
  }
};

export const isFetching = createFetchingReducer([
  FETCH_COMIC_REQUEST,
  FETCH_COMIC_SUCCESS,
  FETCH_COMIC_FAILURE
]);

export const errorMessage = createErrorReducer([
  FETCH_COMIC_REQUEST,
  FETCH_COMIC_SUCCESS,
  FETCH_COMIC_FAILURE
]);

export default combineReducers({
  detail,
  isFetching,
  errorMessage
});

export const getComic = (state, id) => state.detail ? state.detail[id] : state.detail;
export const getIsFetching = state => state.isFetching;
export const getErrorMessage = state => state.errorMessage;
