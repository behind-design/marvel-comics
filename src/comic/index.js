import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';

import { getComic } from '../store/reducers';
import { fetchComic } from '../store/actions';
import ComicDetailComponent from './components/ComicDetail';


export class ComicDetail extends Component {
  componentDidMount() {
    const { comicId, fetchComic } = this.props;

    fetchComic(comicId);
  }

  componentDidUpdate(nextProps) {
    const { fetchComic } = this.props;

    if (nextProps.comicId !== this.props.comicId) {
      return fetchComic(nextProps.comicId);
    }
  }

  render() {
    return <ComicDetailComponent {...this.props} />
  }
}

const mapStateToProps = (state, { match: { params }}) => ({
  comicId: params.id,
  comic: getComic(state, params.id) || {},
});

const mapDispatchToProps = (dispatch) => ({
  fetchComic: (comicId) => dispatch(fetchComic(comicId)),
});

export default withRouter(connect(
  mapStateToProps,
  mapDispatchToProps
)(ComicDetail));
