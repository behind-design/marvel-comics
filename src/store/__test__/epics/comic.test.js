import 'rxjs';
import { Observable } from 'rxjs/Rx';
import { ActionsObservable, createEpicMiddleware } from 'redux-observable';

import * as actions from '../../actions';
import { fetchComicEpic } from '../../epics/comic';


const createAction$ = comicId => ActionsObservable.of(
  { type: actions.FETCH_COMIC_REQUEST, comicId }
)

describe('fetchComicEpic', () => {
  it('should fetch the comic', () => {
    const comicId = 1;
    const results = [
      { id: 1, title: 'Spiderman (1998) #13', prices: [{price: 0}] },
    ];
    const action$ = createAction$(comicId);
    const payload = { response: { data: { results }} };
    const ajax = () => Observable.of(payload);
    const creators = results;
    const characters = results;

    fetchComicEpic(action$, null, { ajax })
      .toArray()
      .subscribe(calledActions => {
        expect(calledActions).toMatchSnapshot();
      });
  });

  it('should call FETCH_COMIC_FAILURE when the request fails', () => {
    const comicId = 1;
    const action$ = createAction$(comicId);
    const error = { message: 'Test error' };
    const ajax = () => Observable.throw(error);

    fetchComicEpic(action$, null, { ajax })
      .toArray()
      .subscribe(calledActions => {
        expect(calledActions).toMatchSnapshot();
      });
  });
});
