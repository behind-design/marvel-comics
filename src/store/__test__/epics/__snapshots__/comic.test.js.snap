// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchComicEpic should call FETCH_COMIC_FAILURE when the request fails 1`] = `
Array [
  Object {
    "comicId": 1,
    "message": "Test error",
    "type": "@@marvel-comics/FETCH_COMIC_FAILURE",
  },
]
`;

exports[`fetchComicEpic should fetch the comic 1`] = `
Array [
  Object {
    "comicId": 1,
    "response": Object {
      "1": Object {
        "characters": Array [
          Object {
            "id": 1,
            "prices": Array [
              Object {
                "price": 0,
              },
            ],
            "title": "Spiderman (1998) #13",
          },
        ],
        "creators": Array [
          Object {
            "id": 1,
            "prices": Array [
              Object {
                "price": 0,
              },
            ],
            "title": "Spiderman (1998) #13",
          },
        ],
        "id": 1,
        "price": "0$",
        "prices": Array [
          Object {
            "price": 0,
          },
        ],
        "thumbnail": "http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available/portrait_uncanny.jpg",
        "title": "Spiderman",
        "year": "1998",
      },
    },
    "type": "@@marvel-comics/FETCH_COMIC_SUCCESS",
  },
]
`;
